File: app/schemas/sms.py
"""Pydantic schemas for SMS inbox and send flows."""

from __future__ import annotations

import uuid
from datetime import datetime

from app.schemas.common import BaseSchema
from app.schemas.org import TwilioBudgetSummary
from app.schemas.voter_contact import PhoneValidationSummary


class SMSEligibilityResponse(BaseSchema):
    """Normalized send-eligibility result for a voter phone."""

    allowed: bool
    reason_code: str | None = None
    reason_detail: str | None = None
    voter_phone_id: uuid.UUID | None = None
    normalized_phone_number: str | None = None
    opt_out_status: str = "active"
    validation: PhoneValidationSummary | None = None


class SMSMessageRead(BaseSchema):
    """Message row rendered in a thread view."""

    id: uuid.UUID
    conversation_id: uuid.UUID
    direction: str
    body: str
    message_type: str
    provider_status: str
    twilio_message_sid: str | None = None
    from_number: str
    to_number: str
    error_code: str | None = None
    error_message: str | None = None
    delivered_at: datetime | None = None
    read_at: datetime | None = None
    created_at: datetime


class SMSConversationRead(BaseSchema):
    """Summary row rendered in the inbox list."""

    id: uuid.UUID
    voter_id: uuid.UUID
    voter_phone_id: uuid.UUID | None = None
    org_phone_number_id: uuid.UUID
    normalized_to_number: str
    last_message_preview: str | None = None
    last_message_direction: str
    last_message_status: str
    last_message_at: datetime | None = None
    unread_count: int
    opt_out_status: str
    opted_out_at: datetime | None = None


class SMSConversationDetail(BaseSchema):
    """Detailed thread payload returned to the UI."""

    conversation: SMSConversationRead
    messages: list[SMSMessageRead]
    eligibility: SMSEligibilityResponse
    budget: TwilioBudgetSummary | None = None


class SMSComposeRequest(BaseSchema):
    """Single-recipient SMS compose request."""

    voter_id: uuid.UUID
    voter_phone_id: uuid.UUID
    body: str


class SMSBulkSendRequest(BaseSchema):
    """Bulk-send request payload."""

    voter_phone_ids: list[uuid.UUID]
    body: str


class SMSBulkSendResponse(BaseSchema):
    """Queued bulk-send response."""

    job_id: str
    queued_count: int
    blocked_count: int = 0
    budget: TwilioBudgetSummary | None = None


class SMSSendResponse(BaseSchema):
    """Single-send response payload."""

    conversation: SMSConversationRead
    message: SMSMessageRead
    eligibility: SMSEligibilityResponse
    budget: TwilioBudgetSummary | None = None


class SMSStatusUpdate(BaseSchema):
    """Delivery status update payload normalized for service usage."""

    twilio_message_sid: str
    provider_status: str
    error_code: str | None = None
    error_message: str | None = None

File: web/src/types/sms.ts
import type { TwilioBudgetSummary } from "@/types/org"
import type { PhoneValidationSummary } from "@/types/voter-contact"

export interface SmsEligibility {
  allowed: boolean
  reason_code?: string | null
  reason_detail?: string | null
  voter_phone_id?: string | null
  normalized_phone_number?: string | null
  opt_out_status: string
  validation?: PhoneValidationSummary | null
}

export interface SmsMessage {
  id: string
  conversation_id: string
  direction: "inbound" | "outbound"
  body: string
  message_type: string
  provider_status: string
  twilio_message_sid?: string | null
  from_number: string
  to_number: string
  error_code?: string | null
  error_message?: string | null
  delivered_at?: string | null
  read_at?: string | null
  created_at: string
}

export interface SmsConversation {
  id: string
  voter_id: string
  voter_phone_id?: string | null
  org_phone_number_id: string
  normalized_to_number: string
  last_message_preview?: string | null
  last_message_direction: "inbound" | "outbound"
  last_message_status: string
  last_message_at?: string | null
  unread_count: number
  opt_out_status: string
  opted_out_at?: string | null
}

export interface SmsConversationDetail {
  conversation: SmsConversation
  messages: SmsMessage[]
  eligibility: SmsEligibility
  budget?: TwilioBudgetSummary | null
}

export interface SmsComposePayload {
  voter_id: string
  voter_phone_id: string
  body: string
}

export interface SmsSendResponse {
  conversation: SmsConversation
  message: SmsMessage
  eligibility: SmsEligibility
  budget?: TwilioBudgetSummary | null
}

export interface SmsBulkSendPayload {
  voter_phone_ids: string[]
  body: string
}

export interface SmsBulkSendResponse {
  job_id: string
  queued_count: number
  blocked_count: number
  budget?: TwilioBudgetSummary | null
}
